compaction.go

```diff
@@ -1680,83 +1672,326 @@ func (d *DB) tryScheduleDownloadCompaction(env compactionEnv, maxConcurrentDownl
 			d.mu.compact.downloads = slices.Delete(d.mu.compact.downloads, i, i+1)
 		}
 	}
-	return false
 }
 
-// maybeScheduleCompactionPicker schedules a compaction if necessary,
-// calling `pickFunc` to pick automatic compactions.
+// makeCompactionEnv attempts to acquire d.mu.versions.logLock in order
+// to provide the proper mutual exclusion necessary during compaction picking.
+// If the DB is closed or marked as read-only, makeCompactionEnv returns nil to
+// indicate that compactions may not be performed. Otherwise, a new
+// compactionEnv is constructed using the current DB state.
+//
+// Compaction picking needs a coherent view of a Version. In particular, we
+// need to exclude concurrent ingestions from making a decision on which level
+// to ingest into that conflicts with our compaction
+// decision. versionSet.logLock provides the necessary mutual exclusion.
+//
+// NOTE: makeCompactionEnv does not call d.mu.versions.logUnlock; it is the
+// caller's responsibility to ensure that the manifest is unlocked.
 //
 // Requires d.mu to be held.
-func (d *DB) maybeScheduleCompactionPicker(
-	pickFunc func(compactionPicker, compactionEnv) *pickedCompaction,
-) {
+func (d *DB) makeCompactionEnv() *compactionEnv {
 	if d.closed.Load() != nil || d.opts.ReadOnly {
-		return
-	}
-	maxCompactions := d.opts.MaxConcurrentCompactions()
-	maxDownloads := d.opts.MaxConcurrentDownloads()
-
-	if d.mu.compact.compactingCount >= maxCompactions &&
-		(len(d.mu.compact.downloads) == 0 || d.mu.compact.downloadingCount >= maxDownloads) {
-		if len(d.mu.compact.manual) > 0 {
-			// Inability to run head blocks later manual compactions.
-			d.mu.compact.manual[0].retries++
-		}
-		return
+		return nil
 	}
 
-	// Compaction picking needs a coherent view of a Version. In particular, we
-	// need to exclude concurrent ingestions from making a decision on which level
-	// to ingest into that conflicts with our compaction
-	// decision. versionSet.logLock provides the necessary mutual exclusion.
 	d.mu.versions.logLock()
```

Comment:
We should just move the responsibility for the logLock to the caller. It's pretty hard to read the compaction pool code without knowing that makeCompactionEnv does the version set locking.